### ISSUE TITLE
abci: Use DeliverTxAsync in execution.go may imply in-deterministic tx execution

### ISSUE THREAD

**Issue by qizhou** [tendermint/tendermint#4529]
https://github.com/tendermint/tendermint/blob/9e5b4afbf1da48d57643b3ccaeb6890b7c16af40/state/execution.go#L293

The code here uses DeliverTxAsync while the rest places use sync versions (e.g., BeginBlockSync, EndBlockSync).  My understanding of using DeliverTxAsync is to speed up tx processing speed.  However, it seems that DeliverTxAsync should be at least waited or replaced by DeliverTxSync due to the following reasons:
- DeliverTxAsync allows delivering multiple txs to the proxy app concurrently, however, if the proxy app processes the txs in parallel (such as multi-threading), the result can be in-deterministic.  If such parallelization is not needed, then using DeliverTxSync will have the same performance result.
- EndBlockSync is sent to proxy app immediately while some DeliverTxAsync requests may still in-flight.  Does it imply that the proxy app must process all in-flight DeliverTxAsync requests before returning EndBlockSync response?

Further, DeliverTxAsync seems to reply on the assumption that the underlying client (grpc or tcp) will implement ordered delivery, which may not be guaranteed if such client implements advanced scheduling such as multi-pathing or QoS.

**Comment by melekes** [comment]
@qizhou thank you for looking into Tendermint! those are great questions 👍 

> My understanding of using DeliverTxAsync is to speed up tx processing speed. 

correct!

"DeliverTx is the workhorse of the blockchain. Tendermint sends the DeliverTx requests asynchronously but in order, and relies on the underlying socket protocol (ie. TCP) to ensure they are received by the app in order. They have already been ordered in the global consensus by the Tendermint protocol."

from https://docs.tendermint.com/master/app-dev/app-development.html#delivertx

**We're sending them in order without waiting for the result (what DeliverTxSync does).**

> EndBlockSync is sent to proxy app immediately while some DeliverTxAsync requests may still in-flight.

no, all requests should be processed by the time app gets EndBlockSync request.

I don't think switching to `DeliverTxSync` will give us anything other than performance hit. The other option we're considering is combing BeginBlock, DeliverTx[] and EndBlock into a single request - DeliverBlock https://github.com/tendermint/tendermint/issues/2901

If you have other questions, feel free to post them here or open another issue. thanks!

**Comment by qizhou** [comment]
Thanks for the clarification.  It makes much more sense if ABCI assumes the underlying proxy app connection (scoket/grpc) delivers DeliverTx in order.

I still have uncertainty about the following statement:

> no, all requests should be processed by the time app gets EndBlockSync request.

From the code, https://github.com/tendermint/tendermint/blob/a89f2581fce4049d8404c7217e8f5c006e14cc52/state/execution.go#L311

I haven't seen any calls such as proxyAppConn.FlushSync(), which bascially waits until all in-flight requests are processed/returned, and thus I am not sure how to guarantee "all requests should be processed by the time app gets EndBlockSync request", i.e., it seems to be possible that an app receives EndBlock request while some DeliverTx requests are delivered but still in-flight in the app due to async nature.

BTW: I like the DeliverBlock idea as all the complicated async races as we discussed above can be deferred to the app.

**Comment by melekes** [comment]
It's certainly not possible if you're using our Go ABCI server because requests are being handled one by one in a simple for loop https://github.com/tendermint/tendermint/blob/a89f2581fce4049d8404c7217e8f5c006e14cc52/abci/server/socket_server.go#L175-L191.

If you're using other servers, it may be possible. Do you know of any such servers? Should we make it more clear in the documentation that introducing parallelism into the ABCI server might lead to a non-determinism? I'm 90% sure ABCI server developers are aware of that fact however.

**Comment by qizhou** [comment]
Thanks for the rely and the code.  Now, I understand how app server handles the ABCI requests.  However, in my opinion, this is an implementation detail, and a good interface should not rely on any implementation details.

Further, using async is naturally implying parallelism, and, in my years of async programming,  I will immediately think about worst possibilities such as
- Out of order delivery
- Out of order processing
which I have seen numerous issues caused by these.

As a result, I think it will be valuable to either await all async requests to be done before any further dependent-request to be issued or use all sync requests.

More question is raised is about the real value of async calls as it implies parallelism and parallelism is not friendly to deterministic replay, a crucial property of blockchain.  One scenario I think it can be a great value is that the blockchain transactions and states are sharded, and the app server processes the transactions in parallel, i.e., using multi-threading with each thread working on one shard.  After all transactions in parallel, EndBlock will summarize all the state hash es from all shards and return an app hash based on those hashes.  Clearly, an await/barrier should be needed to ensure the app hash reflects the results of all processed transcations.
